client/src/main/java/com/netflix/conductor/client/http/ClientBase.java

```diff
@@ -309,121 +219,221 @@ protected Map<String, Object> downloadFromExternalStorage(
-            handleRuntimeException(e, uri);
-        } finally {
-            clientResponse.close();
         }
+
+        return new ConductorClientException(rhe.getMessage(), rhe.getCause());
     }
 
-    private void handleException(URI uri, RuntimeException e) {
-        if (e instanceof UniformInterfaceException) {
-            handleUniformInterfaceException(((UniformInterfaceException) e), uri);
-        } else if (e instanceof ClientHandlerException) {
-            handleClientHandlerException((ClientHandlerException) e, uri);
-        } else {
-            handleRuntimeException(e, uri);
+    private String convertToString(InputStream inputStream) {
+        try {
+            return IOUtils.toString(inputStream, StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            throw new ConductorClientException("Error converting response to String", e);
         }
     }
 
-    /**
-     * Converts ClientResponse object to string with detailed debug information including status
-     * code, media type, response headers, and response body if exists.
-     */
-    private String clientResponseToString(ClientResponse response) {
-        if (response == null) {
-            return null;
+    private <T> T convertToType(InputStream inputStream, Class<T> responseType) {
+        try {
+            return objectMapper.readValue(inputStream, responseType);
+        } catch (IOException e) {
+            throw new ConductorClientException("Error converting response to " + responseType, e);
         }
-        StringBuilder builder = new StringBuilder();
-        builder.append("[status: ").append(response.getStatus());
-        builder.append(", media type: ").append(response.getType());
-        if (response.getStatus() != 404) {
-            try {
-                String responseBody = response.getEntity(String.class);
-                if (responseBody != null) {
-                    builder.append(", response body: ").append(responseBody);
-                }
-            } catch (RuntimeException ignore) {
-                // Ignore if there is no response body, or IO error - it may have already been read
-                // in certain scenario.
-            }
+    }
+
+    private <T> T convertToType(InputStream inputStream, TypeReference<T> responseType) {
+        try {
+            return objectMapper.readValue(inputStream, responseType);
+        } catch (IOException e) {
+            throw new ConductorClientException("Error converting response to " + responseType, e);
         }
-        builder.append(", response headers: ").append(response.getHeaders());
-        builder.append("]");
-        return builder.toString();
     }
 
-    private UriBuilder getURIBuilder(String path, Object[] queryParams) {
-        if (path == null) {
-            path = "";
+    /** An implementation of {@link ExternalPayloadStorage} for storing large JSON payload data. */
+    class PayloadStorage implements ExternalPayloadStorage {
```

Comment:
Can this class still be in its own file?

Reply:
Good catch. It was in its own file and that created a cyclical dependency. `ClientBase` -> `PayloadStorage` -> `ClientBase`. I made it as an inner class to avoid that.

[ClientBase -> PayloadStorage](https://github.com/Netflix/conductor/blob/main/client/src/main/java/com/netflix/conductor/client/http/ClientBase.java#L102)
[PayloadStorage -> ClientBase](https://github.com/Netflix/conductor/blob/main/client/src/main/java/com/netflix/conductor/client/http/PayloadStorage.java#L40)